Fix db history when adding new db

dmaslenko opened the pull request:
Fixes #8375 bug when just created new db was not reflected in the db history (recent files).

Now on a successfully created file it emits a signal to the parent widget to update the last dbs list.

## Testing strategy
### Start the app and check how many recent files you have --> Create a new `test` DB
![welcome-with-one-recent-db](https://user-images.githubusercontent.com/121412908/213900779-3c3d0512-e3c4-45e5-8b19-fc70a303e965.png)

### Close the just created DB
![close-db](https://user-images.githubusercontent.com/121412908/213900796-c95701c8-2498-4886-b4e8-eb1d31d0d2de.png)

### Verify the recent files on the welcome page
![welcome-with-two-recent-db](https://user-images.githubusercontent.com/121412908/213900811-22eb8c87-3aea-4c52-90c7-991ea8b4e4df.png)


### Verify the updated app ini file
![app-config](https://user-images.githubusercontent.com/121412908/213900830-c400bb6f-9a02-44fc-8422-c3b483c47cf0.png)

## Type of change
- ✅ Bug fix (non-breaking change that fixes an issue)

dmaslenko committed:
Fix db history when adding new db

dmaslenko commented:
I am still checking how it can be covered by GUI tests...

droidmonkey commented on src/gui/DatabaseWidget.cpp:
```diff
@@ -216,6 +216,7 @@ DatabaseWidget::DatabaseWidget(QSharedPointer<Database> db, QWidget* parent)
     connect(m_csvImportWizard, SIGNAL(importFinished(bool)), SLOT(csvImportFinished(bool)));
     connect(this, SIGNAL(currentChanged(int)), SLOT(emitCurrentModeChanged()));
     connect(this, SIGNAL(requestGlobalAutoType(const QString&)), parent, SLOT(performGlobalAutoType(const QString&)));
+    connect(this, SIGNAL(updateLastDatabases(const QString&)), parent, SLOT(updateLastDatabases(const QString&)));
```
This is an anti-pattern. You should always connect from parent to child, not child to parent. This should be in DatabaseTabWidget

dmaslenko replied:
I just followed the above example, line 218.
Ok, I moved the `connect` to the `DatabaseTabWidget` class.
Actually this solution does not work, for some reason the `filePath` is empty. May be I missed something?

droidmonkey commented on src/gui/DatabaseWidget.h:
```diff
@@ -153,6 +153,7 @@ class DatabaseWidget : public QStackedWidget
     void clearSearch();
     void requestGlobalAutoType(const QString& search);
     void requestSearch(const QString& search);
+    void updateLastDatabases(const QString& search);
```
No need to introduce a new signal, just use databaseSaved.

dmaslenko replied:
In this case I need to introduce `void updateLastDatabases(int index)` which will call the `void updateLastDatabases(const QString& filePath);` with the current file path.
```c++
void DatabaseTabWidget::updateLastDatabases(int index) {
    if (index == -1) {
        index = currentIndex();
    }

    auto filePath = databaseWidgetFromIndex(index)->database()->filePath();
    // FIXME: the filePath is empty here

    updateLastDatabases(filePath);
}
```
But this solution does not work:
![empty-file-name-in-history](https://user-images.githubusercontent.com/121412908/213905133-f37e79db-5787-4459-a1d7-3ed52a04895a.png)

droidmonkey commented:
It doesn't need gui tests, more trouble than it's worth

dmaslenko committed:
Fix db history when adding new db ver2

droidmonkey commented on src/gui/DatabaseTabWidget.cpp:
```diff
@@ -829,6 +830,17 @@ void DatabaseTabWidget::updateLastDatabases(const QString& filename)
 }
 
+void DatabaseTabWidget::updateLastDatabases(int index) {
+    if (index == -1) {
```
You'll want to check that the index is out of range entirely, so index < 0 || index >= count()

You also need to check that you have a count at all, if there are no database widgets you should just return. 

You can also do this by trying to get the database widget by index, of nullptr then try `currentDatabaseWidget` if nullptr return, otherwise check for a file name and if present continue with the function.

dmaslenko replied:
Thank you for ideas, I will try...

dmaslenko replied:
Check my latest update, I found an example without `index` and as for me it looks more concise. Let me know if a variant with `index` is better.
Also I found a reason why it did not work yesterday: in the `Database` class I changed the order of `setFilePath()` and `markAsClean()`. Now it works.

droidmonkey replied:
That is a good catch with the mark as clean! 

The code is very elegant now, great work

dmaslenko committed:
Simplify updateLastDatabases() to get dbWidget

dmaslenko committed:
Fix saveAs to set the file path before markAsClean

Otherwise the markAsClean emits a signal about saved DB
and for new DB the filePath is still not defined.

dmaslenko committed:
Update last dbs list only if the current file path is not empty

dmaslenko committed:
Add unit test for Database.saveAs()

dmaslenko commented:
I decided to add a unit test for `saveAs()` method to verify the expected signal was emitted and a file was created.

dmaslenko commented:
Wondering why TeamCity builds are pending?

droidmonkey commented:
We have been having a lot of issues with team city controlling the kubernetes based agents

dmaslenko committed:
Add negative test for Database.saveAs() when not initialized db

dmaslenko committed:
Fix failed build on wrong QCOMPARE usage

dmaslenko committed:
Fix formatting issue in DatabaseTabWidget.cpp

dmaslenko committed:
Remove Q_ASSERT in updateLastDatabases() as not required validation

[Why]
  It will fix one CLI test where the widget can be null.
